File: scripts/apply-track-config.jl
using TOML

# Parse config files
global_cfg = TOML.parsefile(joinpath("main", "config.toml"))

# Track config
# Check for existence of track config file
track_cfg_file = joinpath("track-repo", ".github", "org-wide-files-config.toml")
if isfile(track_cfg_file)
    track_cfg = TOML.parsefile(track_cfg_file)

    # Configlet
    if haskey(track_cfg, "configlet")
        # Iterate through allowed configlet settings
        for key in global_cfg["configlet_configurable_keys"]
            haskey(track_cfg["configlet"], key) || continue

            if track_cfg["configlet"]["fmt"]
                @info "Track wants to run configlet fmt; appending to configlet workflow"

                open(joinpath("track-repo", ".github", "workflows", "configlet.yml"), "a") do io
                    write(io, "    with:\n      fmt: true\n")
                end
            end
        end
    end
else
    @info "$(ENV["TRACK"]) does not have an org-wide-files-config.toml"
end

# Appends
# Iterate through files listed in global config file
for file in global_cfg["appendable_files"]
    append_path = joinpath("track-repo", ".appends", file)

    # Check for existance of append file
    isfile(append_path) || (@info "$(ENV["TRACK"]) does not have an append for $file"; continue)

    # Read content of append file
    append_content = read(append_path, String)

    # Read org-wide file
    org_wide_content = readchomp(joinpath("track-repo", file))

    # Append file
    open(joinpath("track-repo", file), "w") do io
        # Write org-wide content
        write(io, org_wide_content)

        # Add or remove trailing newlines depending on global config
        # The default assumption is one trailing newline
        newlines = haskey(global_cfg["files"], file) ? get(global_cfg["files"][file], "trailing_newlines", 1) : 1
        write(io, "\n" ^ newlines)

        # Write append content
        write(io, append_content)

        @info "Appended $file" append_content
    end
end

File: scripts/apply-track-config.py
#!/usr/bin/env python3
"""Apply a target repo's own configuration to the files synced into it.

The sync workflows check this repo out as main/ and the target repo as
track-repo/, then run this script.
"""
import os
import tomllib
from pathlib import Path

MAIN = Path("main")
TARGET = Path("track-repo")
TRACK = os.environ.get("TRACK", "this repo")

with (MAIN / "config.toml").open("rb") as f:
    config = tomllib.load(f)

track_config_file = TARGET / ".github" / "org-wide-files-config.toml"
if track_config_file.is_file():
    with track_config_file.open("rb") as f:
        track_config = tomllib.load(f)

    configlet = track_config.get("configlet", {})
    for key in config["configlet_configurable_keys"]:
        if key not in configlet:
            continue
        if configlet["fmt"]:
            print("Track wants to run configlet fmt; appending to configlet workflow")
            with (TARGET / ".github" / "workflows" / "configlet.yml").open("a") as out:
                out.write("    with:\n      fmt: true\n")
else:
    print(f"{TRACK} does not have an org-wide-files-config.toml")

for name in config["appendable_files"]:
    append_file = TARGET / ".appends" / name
    if not append_file.is_file():
        print(f"{TRACK} does not have an append for {name}")
        continue

    synced = (TARGET / name).read_text().rstrip("\n")
    newlines = config["files"].get(name, {}).get("trailing_newlines", 1)
    (TARGET / name).write_text(synced + "\n" * newlines + append_file.read_text())
    print(f"Appended {name}")
